README.md
# numbamisc
Miscellaneous utilities based on numba and jinja


docs/CHANGES.rst
Changelog for "numbamisc"
-------------------------

Version 0.1.0 (not released)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- initial "stable" release.


tests/test__docs.py
# Built-ins
from __future__ import absolute_import, division, print_function
import doctest

# 3rd party

# This module
import numbamisc


def test_doctests():
    assert doctest.testmod(numbamisc._filters).failed == 0


tests/test_filters.py
# Built-ins
from __future__ import absolute_import, division, print_function

# 3rd party
import numpy as np
import pytest
from scipy import ndimage

# This module
import numbamisc


@pytest.mark.parametrize("kind, cmp_ft",
                         [(numbamisc.median_filter, ndimage.median_filter),
                          (numbamisc.average_filter, None),
                          (numbamisc.sum_filter, None),
                          (numbamisc.median_filter_weigthed, None),
                          (numbamisc.min_filter, ndimage.minimum_filter),
                          (numbamisc.max_filter, ndimage.maximum_filter),
                          ])
@pytest.mark.parametrize("dtype",
                         [np.float_, np.int_])
@pytest.mark.parametrize("ndim",
                         [1, 2, 3, 4, 5])
@pytest.mark.parametrize("border",
                         ['ignore', 'nearest', 'wrap', 'reflect', 'mirror'])
@pytest.mark.parametrize("nan",
                         [False, True])
def test_no_error(kind, dtype, ndim, border, nan, cmp_ft):
    arr = (np.random.random((3, ) * ndim) * 10).astype(dtype)
    mask = np.zeros((3, ) * ndim, bool)
    kernel = np.ones((3, ) * ndim, np.int_)
    res = kind(arr, kernel, mask, mode=border, ignore_nan=nan)
    if border != 'ignore' and not nan:
        if cmp_ft is not None:
            ref = cmp_ft(arr, footprint=kernel, mode=border)
        elif kind is numbamisc.sum_filter:
            ref = ndimage.convolve(arr, kernel, mode=border)
        elif kind is numbamisc.average_filter:
            ref = ndimage.convolve(arr, kernel, mode=border) / np.sum(kernel)
        else:
            return
        np.testing.assert_array_almost_equal(res[0], ref)


numbamisc/_median_filter.py
import numpy as np

from jinja2 import Template
from numba import njit
from scipy.ndimage import median_filter as scipy_median_filter

__all__ = ['median_filter']


ParameterNotSpecified = object()


class MedianFilterTemplate(object):
    """Class to create and return the appropriate median_filter templates."""
    _template = Template("""
def median(image, kernel, mask):
    {% for i in range(0, dimensions) %}
    n{{i}} = image.shape[{{i}}]
    nk{{i}} = kernel.shape[{{i}}]
    wk{{i}} = nk{{i}} // 2
    {% endfor %}
    result = np.zeros(image.shape, dtype=np.float64)
    tmp = np.zeros(kernel.size, dtype=image.dtype)
    {% for i in range(0, dimensions) %}
    {{' ' * i * 4}}for i{{i}} in range(n{{i}}):
    {{' ' * i * 4}}    n{{i}}min = max(i{{i}} - wk{{i}}, 0)
    {{' ' * i * 4}}    n{{i}}max = min(i{{i}} + wk{{i}} + 1, n{{i}})
    {% endfor %}
    {{' ' * dimensions * 4}}elements = 0
    {% for i in range(0, dimensions) %}
    {{' ' * dimensions * 4}}{{' ' * i * 4}}for ii{{i}} in range(n{{i}}min, n{{i}}max):
    {{' ' * dimensions * 4}}{{' ' * i * 4}}    iii{{i}} = wk{{i}} + ii{{i}} - i{{i}}
    {% endfor %}
    {{' ' * dimensions * 8}}if (not mask[{% for i in range(0, dimensions-1) %}ii{{i}}, {% endfor %}ii{{dimensions-1}}] and kernel[{% for i in range(0, dimensions-1) %}iii{{i}}, {% endfor %}iii{{dimensions-1}}]):
    {{' ' * dimensions * 8}}    tmp[elements] = image[{% for i in range(0, dimensions-1) %}ii{{i}}, {% endfor %}ii{{dimensions-1}}]
    {{' ' * dimensions * 8}}    elements += 1

    {{' ' * dimensions * 4}}if elements > 0:
    {{' ' * dimensions * 4}}    result[{% for i in range(0, dimensions-1) %}i{{i}}, {% endfor %}i{{dimensions-1}}] = np.median(tmp[:elements])
    {{' ' * dimensions * 4}}else:
    {{' ' * dimensions * 4}}    result[{% for i in range(0, dimensions-1) %}i{{i}}, {% endfor %}i{{dimensions-1}}] = np.nan
    return result
""")

    _funcs = {}

    def __init__(self):
        pass

    def get_template(self, dimensions):
        if dimensions not in self._funcs:
            exec(self._template.render(dimensions=dimensions), globals())
            # median function is created in the template execution.
            self._funcs[dimensions] = njit(median)
        return self._funcs[dimensions]


def _convert_to_native_bytorder(array):
    """Convert a numpy array with endian dtype to an array of native dtype.

    Parameters
    ----------
    array : `numpy.ndarray`-like
        The array to convert. Must be convertable to a numpy.ndarray.

    Returns
    -------
    converted_array : `numpy.ndarray`
        The converted array. If the array already was in native dtype the array
        is simple returned without making a copy.

    Notes
    -----
    This function is essential when using ``numba`` because it cannot process
    small/big endian arrays. At least it seems like that.
    """
    array = np.asarray(array)
    return array.astype(array.dtype.type, copy=False)


def _process(data, kernel, mask):
    """Checks prerequisites and loads the appropriate function.
    """

    # Try to just extract a mask from the data. If it fails with an
    # AttributeError and no mask was specified create an empty boolean array.
    try:
        mask2 = data.mask
    except AttributeError:
        mask2 = None
    else:
        # In case we sucessfully used the mask of the data we either have a
        # numpy.ma.MaskedArray or similar.
        data = data.data

    data = _convert_to_native_bytorder(data)

    # It is possible that the kernel is an astropy kernel, in that case it has
    # an attribute "array" and we use that one:
    kernel = _convert_to_native_bytorder(getattr(kernel, 'array', kernel))

    # Only in case no explicit mask was given use the one extracted from the
    # data.
    if mask is ParameterNotSpecified:
        mask = mask2

    # In case we have no mask (None) simply use scipy.median_filter
    if mask is None:
        return scipy_median_filter(data, footprint=kernel)

    # Check if the shape is the same. There might be cases where the
    # array contained a mask attribute but the mask has a different shape
    # than the data!
    else:
        mask = np.asarray(mask)
        if data.shape != mask.shape:
            raise ValueError('data and mask must have the same shape.')

    # Evaluate how many dimensions the array has, this is needed to find the
    # appropriate convolution or interpolation function.
    ndim = data.ndim

    # kernel must have the same number of dimensions
    if kernel.ndim != ndim:
        raise ValueError('data and kernel must have the same number of '
                         'dimensions.')

    # the kernel also needs to be odd in every dimension.
    if any(i % 2 == 0 for i in kernel.shape):
        raise ValueError('kernel must have an odd shape in each dimension.')

    func = MedianFilterTemplate().get_template(ndim)

    return func(data, kernel, mask)


def median_filter(data, kernel, mask=ParameterNotSpecified):
    """Median based convolution of some data by ignoring masked values.

    .. note::
        Requires ``numba``.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, :class:`~astropy.nddata.NDData`
        The data to convolve.

    kernel : :class:`~numpy.ndarray`, :class:`~astropy.convolution.Kernel`
        The kernel for the convolution. One difference from normal convolution
        is that the actual values of the kernel do not matter, except when it
        is zero then it won't use the element for the median computation.
        Each axis of the kernel must be odd.

    mask : :class:`~numpy.ndarray`, optional
        Masked values in the ``data``. Elements where the mask is equivalent to
        1 (also ``True``) are interpreted as masked and are ignored during the
        convolution. If not given use the mask of the data or if it has no mask
        either just use `scipy.ndimage.median_filter`.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The median filtered array.

    See also
    --------
    scipy.ndimage.median_filter : Fast n-dimensional convolution \
        without masks.

    Notes
    -----
    If the ``data`` parameter has a ``mask`` attribute then ``data.data``
    is interpreted as ``data`` and ``array.mask`` as ``mask`` parameter.
    This allows using :class:`~numpy.ma.MaskedArray` objects as ``data`` parameter.

    If an explicit ``mask`` is given (even if it is ``None``) an implicit
    mask is ignored.

    No border handling is possible, if the kernel extends beyond the image
    these *outside* values are treated as if they were masked.

    Examples
    --------
    A simple example::

        >>> from numbamisc import median_filter
        >>> import numpy as np

        >>> data = np.ma.array([1,1000,2,1], mask=[0, 1, 0, 0])
        >>> median_filter(data, [1,1,1])
        array([ 1. ,  1.5,  1.5,  1.5])

    Support for arbitarly dimensional arrays and masks is also implemented::

        >>> data = np.arange(9).reshape(3, 3)
        >>> data[1, 1] = 100
        >>> mask = np.zeros((3, 3), dtype=bool)
        >>> mask[1, 1] = 1
        >>> median_filter(data, np.ones((3,3)), mask)
        array([[ 1.,  2.,  2.],
               [ 3.,  4.,  5.],
               [ 6.,  6.,  7.]])

    And another example::

        >>> data = np.arange(27).reshape(3, 3, 3)
        >>> data[0, 0, 0] = 10000
        >>> mask = np.zeros((3, 3, 3))
        >>> mask[0, 0, 0] = 1
        >>> median_filter(data, np.ones((3, 3, 3)), mask)
        array([[[  9. ,   9. ,   7.5],
                [  9. ,   9. ,   9. ],
                [  9.5,  10. ,  10.5]],
        <BLANKLINE>
               [[ 12. ,  12. ,  12. ],
                [ 13. ,  13.5,  13.5],
                [ 14. ,  14.5,  15. ]],
        <BLANKLINE>
               [[ 15.5,  16. ,  16.5],
                [ 17. ,  17.5,  18. ],
                [ 18.5,  19. ,  19.5]]])

    Explictly using kernel elements to zero excludes those elements for the
    convolution::

        >>> data = np.ma.array([1, 1000, 2, 1], mask=[0, 1, 0, 0])
        >>> median_filter(data, [1, 0, 0])
        array([ nan,   1.,  nan,   2.])

    Here only the left element is used for the convolution. For the first
    element the left one is outside the data and for the third element the
    convolution element is masked so both of them result in ``NaN``.
    """
    return _process(data, kernel, mask)


numbamisc/utils/_generatefilters.py
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

import os

if os.environ.get('READTHEDOCS') != 'True':
    import jinja2


__all__ = ['generate']


def stripblanklines(tmp):
    tmp = tmp.split('\n')
    tmp = (i for i in tmp if i.strip())
    return '\n'.join(tmp)


_init = """
import math
import numba as nb
import numpy as np


filters = {}


@nb.njit(nogil=True, cache=True)
def min_ii_ignore(i, nkl, n):
    return max(i - nkl, 0)


@nb.njit(nogil=True, cache=True)
def min_ii_default(i, nkl, n):
    return i - nkl


@nb.njit(nogil=True, cache=True)
def max_ii_ignore(i, nkh, n):
    return min(i + nkh + 1, n)


@nb.njit(nogil=True, cache=True)
def max_ii_default(i, nkh, n):
    return i + nkh + 1


@nb.njit(nogil=True, cache=True)
def img_idx_nearest(ii, n):
    if ii < 0:
        ii = 0
    elif ii >= n:
        ii = n - 1
    return ii


@nb.njit(nogil=True, cache=True)
def img_idx_wrap(ii, n):
    return ii % n


@nb.njit(nogil=True, cache=True)
def img_idx_mirror(ii, n):
    while ii < 0 or ii >= n:
        if ii < 0:
            ii = -ii
        elif ii >= n:
            ii = 2 * n - 2 - ii
    return ii


@nb.njit(nogil=True, cache=True)
def img_idx_reflect(ii, n):
    while ii < 0 or ii >= n:
        if ii < 0:
            ii = -ii - 1
        elif ii >= n:
            ii = 2 * n - 1 - ii
    return ii"""

_convolutiontemplate = """
@nb.njit(nogil=True, cache=True)
def filter_{{mode}}_n{{ndim}}_{{out}}{{nan}}(image, kernel, mask):
    {% for i in range(ndim) %}
    n{{i}} = image.shape[{{i}}]
    nkh{{i}} = kernel.shape[{{i}}] // 2
    nkl{{i}} = nkh{{i}}
    if kernel.shape[{{i}}] % 2 == 0:
        nkh{{i}} -= 1
    {% endfor %}
    resdata = np.zeros(image.shape, dtype=np.float64)
    resmask = np.zeros(image.shape, dtype=np.bool_)

    {% if mode == 'median' %}
    tmp = np.zeros(np.sum(kernel), dtype=image.dtype)
    {% elif mode == 'wmedian' %}
    tmp = np.zeros(np.sum(kernel), dtype=image.dtype)
    {% elif mode == 'sum'%}
    kernelsum = np.sum(kernel)
    {% endif %}

    {% for i in range(ndim) %}
    {% if True             %}{{ ' '*4*i       }}for i{{i}} in range(n{{i}}):{% endif %}
    {% if out == 'ignore'  %}{{ ' '*4*i       }}    n{{i}}min = min_ii_ignore(i{{i}}, nkl{{i}}, n{{i}})
    {% else                %}{{ ' '*4*i       }}    n{{i}}min = min_ii_default(i{{i}}, nkl{{i}}, n{{i}})
    {% endif %}
    {% if out == 'ignore'  %}{{ ' '*4*i       }}    n{{i}}max = max_ii_ignore(i{{i}}, nkh{{i}}, n{{i}})
    {% else                %}{{ ' '*4*i       }}    n{{i}}max = max_ii_default(i{{i}}, nkh{{i}}, n{{i}})
    {% endif %}
    {% endfor %}

    {% if mode == 'min'     %}{{ ' '*4*ndim    }}tmpmin = None{% endif %}
    {% if mode == 'max'     %}{{ ' '*4*ndim    }}tmpmax = None{% endif %}
    {% if mode == 'median'  %}{{ ' '*4*ndim    }}elements = 0{% endif %}
    {% if mode == 'wmedian' %}{{ ' '*4*ndim    }}elements = 0{% endif %}
    {% if mode == 'sum'     %}{{ ' '*4*ndim    }}acc = 0.{% endif %}
    {% if mode == 'sum'     %}{{ ' '*4*ndim    }}div = 0.{% endif %}
    {% if mode == 'mean'    %}{{ ' '*4*ndim    }}acc = 0.{% endif %}
    {% if mode == 'mean'    %}{{ ' '*4*ndim    }}div = 0.{% endif %}

    {% for i in range(ndim) %}
    {% if True             %}{{ ' '*4*(i+ndim) }}for ii{{i}} in range(n{{i}}min, n{{i}}max):{% endif %}
    {% if True             %}{{ ' '*4*(i+ndim) }}    iii{{i}} = nkl{{i}} + ii{{i}} - i{{i}}{% endif %}
    {% if out == 'nearest' %}{{ ' '*4*(i+ndim) }}    ii{{i}} = img_idx_nearest(ii{{i}}, n{{i}}){% endif %}
    {% if out == 'wrap'    %}{{ ' '*4*(i+ndim) }}    ii{{i}} = img_idx_wrap(ii{{i}}, n{{i}}){% endif %}
    {% if out == 'reflect' %}{{ ' '*4*(i+ndim) }}    ii{{i}} = img_idx_reflect(ii{{i}}, n{{i}}){% endif %}
    {% if out == 'mirror'  %}{{ ' '*4*(i+ndim) }}    ii{{i}} = img_idx_mirror(ii{{i}}, n{{i}}){% endif %}
    {% endfor %}

    {{ ' '*8*ndim }}imgitem = image[{% for i in range(ndim-1) %}ii{{i}}, {% endfor %}ii{{ndim-1}}]
    {{ ' '*8*ndim }}maskitem = mask[{% for i in range(ndim-1) %}ii{{i}}, {% endfor %}ii{{ndim-1}}]
    {{ ' '*8*ndim }}kernitem = kernel[{% for i in range(ndim-1) %}iii{{i}}, {% endfor %}iii{{ndim-1}}]
    {{ ' '*8*ndim }}if kernitem and not maskitem{% if nan %} and not math.isnan(imgitem){% endif %}:
    {% if mode == 'median' %}
    {{ ' '*8*ndim }}    tmp[elements] = imgitem
    {{ ' '*8*ndim }}    elements += 1
    {% elif mode == 'min' %}
    {{ ' '*8*ndim }}    if tmpmin is None:
    {{ ' '*8*ndim }}        tmpmin = imgitem
    {{ ' '*8*ndim }}    elif imgitem < tmpmin:
    {{ ' '*8*ndim }}        tmpmin = imgitem
    {% elif mode == 'max' %}
    {{ ' '*8*ndim }}    if tmpmax is None:
    {{ ' '*8*ndim }}        tmpmax = imgitem
    {{ ' '*8*ndim }}    elif imgitem > tmpmax:
    {{ ' '*8*ndim }}        tmpmax = imgitem
    {% elif mode == 'wmedian' %}
    {{ ' '*8*ndim }}    for _ in range(kernitem):
    {{ ' '*8*ndim }}        tmp[elements] = imgitem
    {{ ' '*8*ndim }}        elements += 1
    {% elif mode == 'sum' %}
    {{ ' '*8*ndim }}    acc += imgitem * kernitem
    {{ ' '*8*ndim }}    div += kernitem
    {% elif mode == 'mean' %}
    {{ ' '*8*ndim }}    acc += imgitem * kernitem
    {{ ' '*8*ndim }}    div += kernitem
    {% endif %}

    {% if mode == 'median' %}
    {{ ' '*4*ndim }}if elements != 0:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = np.median(tmp[:elements])
    {% elif mode == 'min' %}
    {{ ' '*4*ndim }}if tmpmin is not None:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = tmpmin
    {% elif mode == 'max' %}
    {{ ' '*4*ndim }}if tmpmax is not None:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = tmpmax
    {% elif mode == 'wmedian' %}
    {{ ' '*4*ndim }}if elements != 0:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = np.median(tmp[:elements])
    {% elif mode == 'sum' %}
    {{ ' '*4*ndim }}if div != 0.:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = acc / div * kernelsum
    {% elif mode == 'mean' %}
    {{ ' '*4*ndim }}if div != 0.:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = acc / div
    {% endif %}
    {{ ' '*4*ndim }}else:
    {{ ' '*4*ndim }}    resdata[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = np.nan
    {{ ' '*4*ndim }}    resmask[{% for i in range(0, ndim-1) %}i{{i}}, {% endfor %}i{{ndim-1}}] = True

    return resdata, resmask

filters[('{{mode}}', {{ndim}}, '{{out}}', bool('{{nan}}'))] = filter_{{mode}}_n{{ndim}}_{{out}}{{nan}}
"""


def generate(maxndim):
    _tmpl = jinja2.Template(_convolutiontemplate)
    _ndim = range(1, maxndim + 1)
    _modes = ['median', 'wmedian', 'sum', 'mean', 'min', 'max']
    _out = ['ignore', 'nearest', 'wrap', 'reflect', 'mirror']
    _nan = ['_ignore_nan', '']

    _funcs = [stripblanklines(_tmpl.render(ndim=ndim, mode=mode,
                                           out=out, nan=nan))
              for mode in _modes
              for ndim in _ndim
              for out in _out
              for nan in _nan]
    _funcs.insert(0, _init)
    return '\n\n\n'.join(_funcs) + '\n'


numbamisc/_filters.py
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

import os

if os.environ.get('READTHEDOCS') != 'True':
    import numpy as np
    from scipy import ndimage as scipy_ndimage

    try:
        from ._genfilters import filters
    except ImportError:
        from .utils import generate

        current_dir = os.path.dirname(__file__)
        target_dir = current_dir

        try:
            os.makedirs(target_dir)
        except OSError:
            pass

        with open(os.path.join(target_dir, '_genfilters.py'), 'w') as fobj:
            fobj.write(generate(maxndim=5))

        from ._genfilters import filters


__all__ = ['median_filter', 'median_filter_weigthed',
           'min_filter', 'max_filter',
           'average_filter', 'sum_filter']


class ParameterNotSpecified(object):
    def __str__(self):
        return "'unspecified'"

    __repr__ = __str__

ParameterNotSpecified = ParameterNotSpecified()


def _convert_to_native_bytorder(array):
    """Convert a numpy array with endian dtype to an array of native dtype.

    Parameters
    ----------
    array : `numpy.ndarray`-like
        The array to convert. Must be convertable to a numpy.ndarray.

    Returns
    -------
    converted_array : `numpy.ndarray`
        The converted array. If the array already was in native dtype the array
        is simple returned without making a copy.

    Notes
    -----
    This function is essential when using ``numba`` because it cannot process
    small/big endian arrays. At least it seems like that.
    """
    array = np.asarray(array)
    return array.astype(array.dtype.type, copy=False)


def _getdatamask(data, mask):
    # Try to just extract a mask from the data. If it fails with an
    # AttributeError and no mask was specified create an empty boolean array.
    try:
        mask2 = data.mask
    except AttributeError:
        mask2 = None
    else:
        # In case we sucessfully used the mask of the data we either have a
        # numpy.ma.MaskedArray or similar.
        data = data.data

    # Only in case no explicit mask was given use the one extracted from the
    # data.
    if mask is ParameterNotSpecified:
        mask = mask2

    data = _convert_to_native_bytorder(data)

    # Check if the shape is the same. There might be cases where the
    # array contained a mask attribute but the mask has a different shape
    # than the data!
    if mask is not None:
        mask = _convert_to_native_bytorder(mask)
        if data.shape != mask.shape:
            raise ValueError('data and mask must have the same shape.')

    return data, mask


def _getkernel(kernel, ndim):
    # Create a kernel from an integer (width in all dimensions)
    if isinstance(kernel, int):
        kernel = np.ones((kernel, ) * ndim, int)

    # Create a kernel from a tuple of integer:
    if isinstance(kernel, tuple):
        kernel = np.ones(kernel, int)

    # It is possible that the kernel is an astropy kernel, in that case it has
    # an attribute "array" and we use that one:
    kernel = _convert_to_native_bytorder(getattr(kernel, 'array', kernel))

    # kernel must have the same number of dimensions
    if kernel.ndim != ndim:
        raise ValueError('data and kernel must have the same number of '
                         'dimensions.')

    return kernel


def _process_input(data, kernel, mask):
    """Checks prerequisites and loads/applies the appropriate function.
    """
    data, mask = _getdatamask(data, mask)

    kernel = _getkernel(kernel, data.ndim)

    return data, kernel, mask


def median_filter(data, kernel, mask=ParameterNotSpecified,
                  mode='ignore', ignore_nan=True):
    """Median filter ignoring masked and NaN values.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, \
:class:`~astropy.nddata.NDData`
        The data to filter.

    kernel : :class:`int`, :class:`tuple`, :class:`~numpy.ndarray`, \
:class:`~astropy.convolution.Kernel`
        See main documentation for explanation.
        The ``kernel`` is cast to ``dtype=bool`` before filtering.

    mask : :class:`~numpy.ndarray`, optional
        Mask for the filter. If given an implicit mask by ``data`` is ignored.

    mode : string, optional
        How to treat values outside the ``data``.
        Default is ``ignore``.

    ignore_nan : bool, optional
        Also ignore ``NaN`` values.
        Default is ``True``.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The filtered array.

    mask : :class:`~numpy.ndarray`
        The mask of the filtered array.

    See also
    --------
    scipy.ndimage.median_filter : the same without mask support.

    Examples
    --------
    A simple example using a masked array::

        >>> from numbamisc import median_filter
        >>> import numpy as np

        >>> data = np.ma.array([1,1000,2,1], mask=[0, 1, 0, 0])
        >>> d, m = median_filter(data, [1,1,1])
        >>> d
        array([ 1. ,  1.5,  1.5,  1.5])
        >>> m
        array([False, False, False, False], dtype=bool)

    Support for arbitarly dimensional arrays and masks is also implemented::

        >>> data = np.arange(9).reshape(3, 3)
        >>> data[1, 1] = 100
        >>> mask = np.zeros((3, 3), dtype=bool)
        >>> mask[1, 1] = 1
        >>> d, m = median_filter(data, np.ones((3,3)), mask)
        >>> d
        array([[ 1.,  2.,  2.],
               [ 3.,  4.,  5.],
               [ 6.,  6.,  7.]])
        >>> m
        array([[False, False, False],
               [False, False, False],
               [False, False, False]], dtype=bool)

    Explictly setting kernel elements to zero excludes those elements for the
    convolution::

        >>> data = np.ma.array([1, 1000, 2, 1], mask=[0, 1, 0, 0])
        >>> d, m = median_filter(data, [1, 0, 0])
        >>> d
        array([ nan,   1.,  nan,   2.])
        >>> m
        array([ True, False,  True, False], dtype=bool)

    Here only the left element is used for the convolution. For the first
    element the left one is outside the data and for the third element the
    convolution element is masked so both of them result in ``NaN``.
    """
    data, kernel, mask = _process_input(data, kernel, mask)

    # In case we have no mask (None) simply use scipy.median_filter
    if mask is None:
        return scipy_ndimage.median_filter(data, footprint=kernel)

    kernel = kernel.astype(bool)

    func = filters[('median', data.ndim, mode, ignore_nan)]

    return func(data, kernel, mask)


def median_filter_weigthed(data, kernel, mask=ParameterNotSpecified,
                           mode='ignore', ignore_nan=True):
    """Weighted median filter ignoring masked and NaN values.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, \
:class:`~astropy.nddata.NDData`
        The data to filter.

    kernel : :class:`int`, :class:`tuple`, :class:`~numpy.ndarray`, \
:class:`~astropy.convolution.Kernel`
        See main documentation for explanation.
        The ``kernel`` is cast to ``dtype=np.int_`` before filtering.

    mask : :class:`~numpy.ndarray`, optional
        Mask for the filter. If given an implicit mask by ``data`` is ignored.

    mode : string, optional
        How to treat values outside the ``data``.
        Default is ``ignore``.

    ignore_nan : bool, optional
        Also ignore ``NaN`` values.
        Default is ``True``.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The filtered array.

    mask : :class:`~numpy.ndarray`
        The mask of the filtered array.
    """
    data, kernel, mask = _process_input(data, kernel, mask)

    if mask is None:
        mask = np.zeros(data.shape, dtype=bool)

    kernel = kernel.astype(np.int_)

    func = filters[('wmedian', data.ndim, mode, ignore_nan)]

    return func(data, kernel, mask)


def min_filter(data, kernel, mask=ParameterNotSpecified,
               mode='ignore', ignore_nan=True):
    """Minimum filter ignoring masked and NaN values.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, \
:class:`~astropy.nddata.NDData`
        The data to filter.

    kernel : :class:`int`, :class:`tuple`, :class:`~numpy.ndarray`, \
:class:`~astropy.convolution.Kernel`
        See main documentation for explanation.
        The ``kernel`` is cast to ``dtype=bool`` before filtering.

    mask : :class:`~numpy.ndarray`, optional
        Mask for the filter. If given an implicit mask by ``data`` is ignored.

    mode : string, optional
        How to treat values outside the ``data``.
        Default is ``ignore``.

    ignore_nan : bool, optional
        Also ignore ``NaN`` values.
        Default is ``True``.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The median filtered array.

    mask : :class:`~numpy.ndarray`
        The mask of the filtered array.

    See also
    --------
    scipy.ndimage.minimum_filter : the same without mask support.
    """
    data, kernel, mask = _process_input(data, kernel, mask)

    if mask is None:
        mask = np.zeros(data.shape, dtype=bool)

    kernel = kernel.astype(bool)

    func = filters[('min', data.ndim, mode, ignore_nan)]

    return func(data, kernel, mask)


def max_filter(data, kernel, mask=ParameterNotSpecified,
               mode='ignore', ignore_nan=True):
    """Maximum filter ignoring masked and NaN values.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, \
:class:`~astropy.nddata.NDData`
        The data to filter.

    kernel : :class:`int`, :class:`tuple`, :class:`~numpy.ndarray`, \
:class:`~astropy.convolution.Kernel`
        See main documentation for explanation.
        The ``kernel`` is cast to ``dtype=bool`` before filtering.

    mask : :class:`~numpy.ndarray`, optional
        Mask for the filter. If given an implicit mask by ``data`` is ignored.

    mode : string, optional
        How to treat values outside the ``data``.
        Default is ``ignore``.

    ignore_nan : bool, optional
        Also ignore ``NaN`` values.
        Default is ``True``.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The median filtered array.

    mask : :class:`~numpy.ndarray`
        The mask of the filtered array.

    See also
    --------
    scipy.ndimage.maximum_filter : the same without mask support.
    """
    data, kernel, mask = _process_input(data, kernel, mask)

    if mask is None:
        mask = np.zeros(data.shape, dtype=bool)

    kernel = kernel.astype(bool)

    func = filters[('max', data.ndim, mode, ignore_nan)]

    return func(data, kernel, mask)


def sum_filter(data, kernel, mask=ParameterNotSpecified,
               mode='ignore', ignore_nan=True):
    """Summation filter ignoring masked and NaN values.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, \
:class:`~astropy.nddata.NDData`
        The data to filter.

    kernel : :class:`int`, :class:`tuple`, :class:`~numpy.ndarray`, \
:class:`~astropy.convolution.Kernel`
        See main documentation for explanation.
        The kernel must not contain mixed signed values. Either all values
        must be positive or all negative.

    mask : :class:`~numpy.ndarray`, optional
        Mask for the filter. If given an implicit mask by ``data`` is ignored.

    mode : string, optional
        How to treat values outside the ``data``.
        Default is ``ignore``.

    ignore_nan : bool, optional
        Also ignore ``NaN`` values.
        Default is ``True``.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The median filtered array.

    mask : :class:`~numpy.ndarray`
        The mask of the filtered array.

    See also
    --------
    scipy.ndimage.convolve : the same without mask support.
    """
    data, kernel, mask = _process_input(data, kernel, mask)

    if mask is None:
        mask = np.zeros(data.shape, dtype=bool)

    func = filters[('sum', data.ndim, mode, ignore_nan)]

    return func(data, kernel, mask)


def average_filter(data, kernel, mask=ParameterNotSpecified,
                   mode='ignore', ignore_nan=True):
    """Averaging filter ignoring masked and NaN values.

    Parameters
    ----------
    data : :class:`~numpy.ndarray`, :class:`~numpy.ma.MaskedArray`, \
:class:`~astropy.nddata.NDData`
        The data to filter.

    kernel : :class:`int`, :class:`tuple`, :class:`~numpy.ndarray`, \
:class:`~astropy.convolution.Kernel`
        See main documentation for explanation.
        The kernel must not contain mixed signed values. Either all values
        must be positive or all negative.

    mask : :class:`~numpy.ndarray`, optional
        Mask for the filter. If given an implicit mask by ``data`` is ignored.

    mode : string, optional
        How to treat values outside the ``data``.
        Default is ``ignore``.

    ignore_nan : bool, optional
        Also ignore ``NaN`` values.
        Default is ``True``.

    Returns
    -------
    filtered : :class:`~numpy.ndarray`
        The median filtered array.

    mask : :class:`~numpy.ndarray`
        The mask of the filtered array.
    """
    data, kernel, mask = _process_input(data, kernel, mask)

    if mask is None:
        mask = np.zeros(data.shape, dtype=bool)

    func = filters[('mean', data.ndim, mode, ignore_nan)]

    return func(data, kernel, mask)
